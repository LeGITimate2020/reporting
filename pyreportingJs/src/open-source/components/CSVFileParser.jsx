import React from 'react';
import PropTypes from 'prop-types';
import { DropzoneComponent } from 'react-dropzone-component';
import Papa from 'papaparse';


/**
 * CSVFileParser allows a user to drag-and-drop and copy & paste a CSV file and then it parses it into an Object and
 * calls an event handler supplied by the parent component.
 * All processing is done in-browser and data is never uploaded or sent to a server.
 */
export default class CSVFileParser extends React.Component {
  static propTypes = {
    onLoad: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {};
    this.dropzone = null;
  }

  handleDropzoneInit = (dropzone) => {
    this.dropzone = dropzone;
  };

  handleParseCSV = fileName => (evt) => {
    const rawFile = evt.target.result;
    const csvResults = Papa.parse(rawFile);
    this.props.onLoad(csvResults, fileName);
  };

  handleAddedFile = (addedFile) => {
    if (addedFile) {
      const fileName = addedFile.upload.filename;
      const reader = new FileReader();
      reader.onload = this.handleParseCSV(fileName);
      reader.readAsText(addedFile, 'UTF-8');

      // Remove file from preview component
      this.dropzone.removeFile(addedFile);
    }
  };

  render() {
    const componentConfig = {
      iconFiletypes: ['.csv', '.tsv'],
      showFiletypeIcon: true,
      postUrl: '/',
    };

    // Full list of supported eventHandlers: https://github.com/felixrieseberg/React-Dropzone-Component
    const eventHandlers = {
      init: this.handleDropzoneInit,
      addedfile: this.handleAddedFile,
    };

    const djsConfig = {
      autoProcessQueue: false,
      dictDefaultMessage: 'Drag and drop a .CSV file here to use as your data source',
    };

    return (
      <div className="row">
        <div className="col-lg-12">
          <DropzoneComponent
            config={componentConfig}
            action="api/files/upload"
            eventHandlers={eventHandlers}
            djsConfig={djsConfig}
          />
        </div>
      </div>
    );
  }
}
