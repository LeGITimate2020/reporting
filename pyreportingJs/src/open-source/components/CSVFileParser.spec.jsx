import { shallow } from 'enzyme';
import React from 'react';
import { DropzoneComponent } from 'react-dropzone-component';
import CSVFileParser from './CSVFileParser';

describe('CSVFileParser.jsx', () => {
  const noOp = () => {};
  const props = {
    onLoad: noOp,
  };

  it('renders without crashing', () => {
    shallow(<CSVFileParser {...props} />);
  });

  it('Return DropzoneComponent with correct props', () => {
    const wrapper = shallow(<CSVFileParser {...props} />);
    // Should be 1 Dropzonecomponent
    expect(wrapper.find(DropzoneComponent).length).toBe(1);
  });

  it('Matches snapshot', () => {
    const wrapper = shallow(<CSVFileParser {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
