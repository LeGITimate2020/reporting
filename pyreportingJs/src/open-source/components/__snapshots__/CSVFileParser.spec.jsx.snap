// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSVFileParser.jsx Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CSVFileParser
    onLoad={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="col-lg-12"
      >
        <t
          action="api/files/upload"
          config={
            Object {
              "iconFiletypes": Array [
                ".csv",
                ".tsv",
              ],
              "postUrl": "/",
              "showFiletypeIcon": true,
            }
          }
          djsConfig={
            Object {
              "autoProcessQueue": false,
              "dictDefaultMessage": "Drag and drop a .CSV file here to use as your data source",
            }
          }
          eventHandlers={
            Object {
              "addedfile": [Function],
              "init": [Function],
            }
          }
        />
      </div>,
      "className": "row",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <t
          action="api/files/upload"
          config={
            Object {
              "iconFiletypes": Array [
                ".csv",
                ".tsv",
              ],
              "postUrl": "/",
              "showFiletypeIcon": true,
            }
          }
          djsConfig={
            Object {
              "autoProcessQueue": false,
              "dictDefaultMessage": "Drag and drop a .CSV file here to use as your data source",
            }
          }
          eventHandlers={
            Object {
              "addedfile": [Function],
              "init": [Function],
            }
          }
        />,
        "className": "col-lg-12",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": "api/files/upload",
          "config": Object {
            "iconFiletypes": Array [
              ".csv",
              ".tsv",
            ],
            "postUrl": "/",
            "showFiletypeIcon": true,
          },
          "djsConfig": Object {
            "autoProcessQueue": false,
            "dictDefaultMessage": "Drag and drop a .CSV file here to use as your data source",
          },
          "eventHandlers": Object {
            "addedfile": [Function],
            "init": [Function],
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="col-lg-12"
        >
          <t
            action="api/files/upload"
            config={
              Object {
                "iconFiletypes": Array [
                  ".csv",
                  ".tsv",
                ],
                "postUrl": "/",
                "showFiletypeIcon": true,
              }
            }
            djsConfig={
              Object {
                "autoProcessQueue": false,
                "dictDefaultMessage": "Drag and drop a .CSV file here to use as your data source",
              }
            }
            eventHandlers={
              Object {
                "addedfile": [Function],
                "init": [Function],
              }
            }
          />
        </div>,
        "className": "row",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <t
            action="api/files/upload"
            config={
              Object {
                "iconFiletypes": Array [
                  ".csv",
                  ".tsv",
                ],
                "postUrl": "/",
                "showFiletypeIcon": true,
              }
            }
            djsConfig={
              Object {
                "autoProcessQueue": false,
                "dictDefaultMessage": "Drag and drop a .CSV file here to use as your data source",
              }
            }
            eventHandlers={
              Object {
                "addedfile": [Function],
                "init": [Function],
              }
            }
          />,
          "className": "col-lg-12",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": "api/files/upload",
            "config": Object {
              "iconFiletypes": Array [
                ".csv",
                ".tsv",
              ],
              "postUrl": "/",
              "showFiletypeIcon": true,
            },
            "djsConfig": Object {
              "autoProcessQueue": false,
              "dictDefaultMessage": "Drag and drop a .CSV file here to use as your data source",
            },
            "eventHandlers": Object {
              "addedfile": [Function],
              "init": [Function],
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
